kong/plugins/ai-prompt-template/handler.lua

```diff
@@ -0,0 +1,131 @@
+--
+-- @local
+-- @function is_reference
+-- @tparam string reference reference to check
+-- @treturn boolean `true` is the passed in reference looks like a reference, otherwise `false`
+local function is_reference(reference)
+  return type(reference)      == "string"
+     and byte(reference, 1)   == BRACE_START
+     and byte(reference, -1)  == BRACE_END
+     and byte(reference, 10)   == COLON
+     and byte(reference, 11)   == SLASH
+     and byte(reference, 12)   == SLASH
+     and sub(reference, 2, 9) == "template"
+end
+
+local function find_template(reference_string, templates)
+  if is_reference(reference_string) then
+    local parts, err = parse_url(sub(reference_string, 2, -2))
+    if not parts then
+      return nil, fmt("template reference is not in format '{template://template_name}' (%s) [%s]", err, reference_string)
+    end
+
+    -- iterate templates to find it
+    for i, v in ipairs(templates) do
+      if v.name == parts.host then
+        return v, nil
+      end
+    end
+
+    return nil, "could not find template name [" .. parts.host .. "]"
+  else
+    return nil, "'messages' template reference should be a single string, in format '{template://template_name}'"
+  end
+end
+
+function _M:access(conf)
+  kong.service.request.enable_buffering()
+  kong.ctx.shared.ai_prompt_templated = true
+
+  if conf.log_original_request then
+    kong.log.set_serialize_value(log_entry_keys.REQUEST_BODY, kong.request.get_raw_body())
+  end
+
+  -- if plugin ordering was altered from a previous AI-family plugin, use the replacement request
+  local request, err
+  if not kong.ctx.replacement_request then
+    request, err = kong.request.get_body("application/json")
+
+    if err then
+      return bad_request("ai-prompt-template only supports application/json requests")
+    end
+  else
+    request = kong.ctx.replacement_request
+  end
+
+  if (not request.messages) and (not request.prompt) then
+    return bad_request("ai-prompt-template only support llm/chat or llm/completions type requests")
+  end
+
+  if request.messages and request.prompt then
+    return bad_request("cannot run 'messages' and 'prompt' templates at the same time")
+  end
+
+  local reference
+  if request.messages then
+    reference = request.messages
+
+  elseif request.prompt then
+    reference = request.prompt
+
+  else
+    return bad_request("only 'llm/v1/chat' and 'llm/v1/completions' formats are supported for templating")
+  end
+
+  local requested_template, err = find_template(reference, conf.templates)
+  if err and (not conf.allow_untemplated_requests) then bad_request(err) end
+
+  if not err then
+    -- try to render the replacement request
+    local rendered_template, err = templater:render(requested_template, request.properties or {})
+    if err then return bad_request(err) end
+
+    local result, err = cjson.decode(rendered_template)
+    if err then bad_request("failed to parse template to JSON: " .. err) end
```

Comment:
I think this `err` will not be catched because we are using plain `cjson` module. Please use `cjson.safe` if you want to handle JSON errors. The `cjson.safe` module behaves identically to the cjson module, except when errors are encountered during JSON conversion. On error, the cjson_safe.encode and cjson_safe.decode functions will return nil followed by the error message. Ref: https://www.kyne.com.au/~mark/software/lua-cjson-manual.html

I also suggest to add a test case for this.

Reply:
It is using cjson.safe